### ISSUE TITLE
Replacing `state` on the server breaks the state on the client

### ISSUE THREAD

**Issue by Byloth** [colyseus/colyseus-construct3#32]
I'm implementing a game like Fall Guys, where I used to replace the `state` of the room as I transitioned from a minigame to another.  
Each room state has its own properties and logic. It seemed to work fine.

Then, I started adding new complex properties (mainly `Map` properties) and something went wrong.
In particular, I had some issues with a couple of `Map` properties that were defined in both the previous state and the new state.

I used to initialize the new state by assigning the previous cloned property, as follows:

```ts
export class PlayerState extends Schema
{
  @type("string")
  public username: string;

  @type("string")
  public skin: string;

  [...]
}
export class PositionState extends Schema
{
  @type("number")
  public x: number;

  @type("number")
  public y: number;

  [...]
}

export class NewState extends Schema
{
  @type({ map: PlayerState })
  public players: MapSchema<PlayerState>

  @type({ map: PositionState })
  public positions: MapSchema<PositionState>

  public constructor(state: PreviousState)
  {
    this.players = state.players.clone()
    this.positions = state.positions.clone()
  }
}
```

After switching from one minigame to the next, replacing the state from the previous to the new one and once sent to the client, I ended up with a mixed state, somehow weird... I found `username` / `skin` properties inside `positions` map for some players and `x` / `y` inside `players` for some others...

Now... I'm not sure where the problem actually is...  
I would say, not on the server-side since every time I logged those properties on the server-side, they always seemed to be fine...

I guess the issue is on the client-side... Maybe where state binary updates are handled...  
Not sure if this problem appears only with Construct 3 plugin or also with [Colyseus.js](https://github.com/colyseus/colyseus.js) in general... 🤔

**Comment by endel** [comment]
Hi @Byloth, I would suggest to never replace the root "state"  instance (either via `this.state=...` or `this.setState()`) on the server. It's recommended to clear its structures instead. 

Swapping the root state to a different type is strictly not allowed due to how the "handshake" works when first connecting into the room. The client-side only receives the structures of a single state type when joining, so it only knows how to build 1 state type. Swapping will cause decoding issues

If you believe you found a bug I'll kindly ask for a small server + client implementation that reproduces the bug 🙏  

Thank you!

**Comment by Byloth** [comment]
Hi, @endel! 🙃

No, ok... If it's something already known, for me it's ok.  
It could be useful -maybe- to report that you cannot swap the `state` to a different type in the documentation too. 🤔

... or -yet better- implement some sort of control on `setState`:

```js
  setState(newState) {
    if (this.state) {
        throw new Error("State has already been set. You cannot replace it.");
    }
    
    // ... or if you want to make sure to only accept new states strictly of the same class...
    
    if (newState.constructor !== this.state.constructor) {
        throw new Error("State cannot be swapped with an object of a different type.");
    }

    // [...]
  }
```

**Comment by endel** [comment]
Thank you for the suggestion! The `state` getter/setter is already implemented on `0.16` ☺️ 

Adding this new check is a good idea!
